File: sdk/go/pkg/artifact/step.go
package artifact

import (
	"bytes"
	"fmt"
	"strings"
	"text/template"

	api "github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/api/artifact"
	"github.com/ALT-F4-LLC/vorpal/sdk/go/pkg/config"
)

type BashScriptTemplateArgs struct {
	Script string
}

type BwrapScriptTemplateArgs struct {
	Script string
}

const BashScriptTemplate = `#!/bin/bash
set -euo pipefail
{{.Script}}
`

const BwrapScriptTemplate = `#!/bin/bash
set -euo pipefail
{{.Script}}
`

func Bash(
	artifacts []*string,
	environments []string,
	script string,
	secrets []*api.ArtifactStepSecret,
) (*api.ArtifactStep, error) {
	stepEnvironments := make([]string, 0)

	stepEntrypoint := "bash"

	for _, value := range environments {
		if strings.HasPrefix(value, "PATH=") {
			continue
		}

		stepEnvironments = append(stepEnvironments, value)
	}

	stepPathBins := make([]string, 0)

	for _, art := range artifacts {
		if art != nil {
			stepPathBins = append(stepPathBins, fmt.Sprintf("%s/bin", GetEnvKey(*art)))
		}
	}

	stepPathDefault := "/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin"

	stepPath := fmt.Sprintf("%s:%s", strings.Join(stepPathBins, ":"), stepPathDefault)

	for _, value := range environments {
		if pathValue, ok := strings.CutPrefix(value, "PATH="); ok {
			stepPath = fmt.Sprintf("%s:%s", pathValue, stepPath)
		}
	}

	stepEnvironments = append(stepEnvironments, "HOME=$VORPAL_WORKSPACE")
	stepEnvironments = append(stepEnvironments, fmt.Sprintf("PATH=%s", stepPath))

	scriptTemplate, err := template.New("script").Parse(BashScriptTemplate)
	if err != nil {
		return nil, err
	}

	var scriptBuffer bytes.Buffer

	scriptTemplateVars := BashScriptTemplateArgs{
		Script: script,
	}

	if err := scriptTemplate.Execute(&scriptBuffer, scriptTemplateVars); err != nil {
		return nil, err
	}

	step := NewArtifactStep(stepEntrypoint)

	step = step.WithArtifacts(artifacts)
	step = step.WithEnvironments(stepEnvironments)
	step = step.WithScript(scriptBuffer.String())
	step = step.WithSecrets(secrets)

	return step.Build(), nil
}

func Bwrap(
	arguments []string,
	artifacts []*string,
	environments []string,
	rootfs *string,
	script string,
	secrets []*api.ArtifactStepSecret,
) (*api.ArtifactStep, error) {
	// Setup arguments

	stepArguments := []string{
		"--unshare-all",
		"--share-net",
		"--clearenv",
		"--chdir",
		"$VORPAL_WORKSPACE",
		"--gid",
		"1000",
		"--uid",
		"1000",
		"--dev",
		"/dev",
		"--proc",
		"/proc",
		"--tmpfs",
		"/tmp",
		"--bind",
		"$VORPAL_OUTPUT",
		"$VORPAL_OUTPUT",
		"--bind",
		"$VORPAL_WORKSPACE",
		"$VORPAL_WORKSPACE",
		"--setenv",
		"VORPAL_OUTPUT",
		"$VORPAL_OUTPUT",
		"--setenv",
		"VORPAL_WORKSPACE",
		"$VORPAL_WORKSPACE",
		"--setenv",
		"HOME",
		"$VORPAL_WORKSPACE",
	}

	// Setup artifacts arguments

	stepArtifacts := make([]*string, 0)

	if rootfs != nil {
		rootfsArgs := []string{
			"--ro-bind",
			fmt.Sprintf("%s/bin", GetEnvKey(*rootfs)),
			"/bin",
			"--ro-bind",
			fmt.Sprintf("%s/etc", GetEnvKey(*rootfs)),
			"/etc",
			"--ro-bind",
			fmt.Sprintf("%s/lib", GetEnvKey(*rootfs)),
			"/lib",
			"--ro-bind-try",
			fmt.Sprintf("%s/lib64", GetEnvKey(*rootfs)),
			"/lib64",
			"--ro-bind",
			fmt.Sprintf("%s/sbin", GetEnvKey(*rootfs)),
			"/sbin",
			"--ro-bind",
			fmt.Sprintf("%s/usr", GetEnvKey(*rootfs)),
			"/usr",
		}

		stepArguments = append(stepArguments, rootfsArgs...)
		stepArtifacts = append(stepArtifacts, rootfs)
	}

	for _, artifact := range artifacts {
		stepArtifacts = append(stepArtifacts, artifact)
	}

	for _, art := range stepArtifacts {
		if art != nil {
			stepArguments = append(stepArguments, "--ro-bind")
			stepArguments = append(stepArguments, GetEnvKey(*art))
			stepArguments = append(stepArguments, GetEnvKey(*art))
			stepArguments = append(stepArguments, "--setenv")
			stepArguments = append(stepArguments, strings.ReplaceAll(GetEnvKey(*art), "$", ""))
			stepArguments = append(stepArguments, GetEnvKey(*art))
		}
	}

	// Setup environment arguments

	stepPathBins := make([]string, 0)

	for _, art := range stepArtifacts {
		if art != nil {
			stepPathBins = append(stepPathBins, fmt.Sprintf("%s/bin", GetEnvKey(*art)))
		}
	}

	stepPath := fmt.Sprintf("%s:%s", strings.Join(stepPathBins, ":"), "/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin")

	for _, envvar := range environments {
		if pathValue, ok := strings.CutPrefix(envvar, "PATH="); ok {
			stepPath = fmt.Sprintf("%s:%s", pathValue, stepPath)
		}
	}

	stepArguments = append(stepArguments, "--setenv")
	stepArguments = append(stepArguments, "PATH")
	stepArguments = append(stepArguments, stepPath)

	for _, envvar := range environments {
		parts := strings.SplitN(envvar, "=", 2)
		if len(parts) != 2 {
			continue
		}

		key := parts[0]
		value := parts[1]

		if key == "PATH" {
			continue
		}

		stepArguments = append(stepArguments, "--setenv")
		stepArguments = append(stepArguments, key)
		stepArguments = append(stepArguments, value)
	}

	// Setup arguments

	for _, argument := range arguments {
		stepArguments = append(stepArguments, argument)
	}

	// Setup script

	scriptTemplate, err := template.New("script").Parse(BwrapScriptTemplate)
	if err != nil {
		return nil, err
	}

	var scriptBuffer bytes.Buffer

	scriptTemplateVars := BwrapScriptTemplateArgs{
		Script: script,
	}

	if err := scriptTemplate.Execute(&scriptBuffer, scriptTemplateVars); err != nil {
		return nil, err
	}

	// Setup step

	step := NewArtifactStep("bwrap")

	step = step.WithArguments(stepArguments)
	step = step.WithArtifacts(stepArtifacts)
	step = step.WithEnvironments([]string{"PATH=/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin"})
	step = step.WithScript(scriptBuffer.String())
	step = step.WithSecrets(secrets)

	return step.Build(), nil
}

func Shell(
	context *config.ConfigContext,
	artifacts []*string,
	environments []string,
	script string,
	secrets []*api.ArtifactStepSecret,
) (*api.ArtifactStep, error) {
	stepSystem := context.GetTarget()

	if stepSystem == api.ArtifactSystem_AARCH64_DARWIN || stepSystem == api.ArtifactSystem_X8664_DARWIN {
		return Bash(
			artifacts,
			environments,
			script,
			secrets,
		)
	}

	if stepSystem == api.ArtifactSystem_AARCH64_LINUX || stepSystem == api.ArtifactSystem_X8664_LINUX {
		linux_vorpal, err := context.FetchArtifactAlias("library/linux-vorpal:latest")
		if err != nil {
			return nil, err
		}

		return Bwrap(
			[]string{},
			artifacts,
			environments,
			linux_vorpal,
			script,
			secrets,
		)
	}

	return nil, fmt.Errorf("unsupported shell step system: %s", stepSystem)
}

// TODO: Add support for secrets with docker step

func Docker(
	arguments []string,
	artifacts []*string,
) (*api.ArtifactStep, error) {
	step := NewArtifactStep("docker")

	step = step.WithArguments(arguments)
	step = step.WithArtifacts(artifacts)
	step = step.WithEnvironments([]string{"PATH=/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin"})

	return step.Build(), nil
}

File: sdk/python/src/vorpal_sdk/step.py
"""Step builders (``bash``/``bwrap``/``shell``/``docker``).

Mirrors ``sdk/typescript/src/artifact/step.ts`` (the canonical reference) and
``sdk/go/pkg/artifact/step.go``. Sync model — no async, unlike the TS SDK.

These are pure functions of their inputs except ``shell``, which dispatches on
the target system via ``BuildContext.get_system``. The Linux branch needs the
``linux-vorpal`` rootfs builder (a later phase); until it lands, ``shell``
raises on Linux targets. The Darwin branch (``bash``) is fully implemented.
"""

from __future__ import annotations

from typing import Protocol, runtime_checkable

from vorpal_sdk.api.artifact import artifact_pb2

_DEFAULT_PATH = "/usr/local/bin:/usr/bin:/usr/sbin:/bin:/sbin"


@runtime_checkable
class BuildContext(Protocol):
    """The slice of ``ConfigContext`` (Phase 5) the builders depend on.

    Defining it as a Protocol lets Phase-4 builders type-check against the
    not-yet-implemented context and lets tests supply a lightweight stub.
    """

    def get_system(self) -> artifact_pb2.ArtifactSystem: ...

    def add_artifact(self, artifact: artifact_pb2.Artifact) -> str: ...

    def get_variable(self, name: str) -> str | None: ...

    def get_artifact_namespace(self) -> str: ...


def _dedupe_secrets(
    secrets: list[artifact_pb2.ArtifactStepSecret],
) -> list[artifact_pb2.ArtifactStepSecret]:
    seen: set[str] = set()
    unique: list[artifact_pb2.ArtifactStepSecret] = []
    for secret in secrets:
        if secret.name not in seen:
            seen.add(secret.name)
            unique.append(secret)
    return unique


def bash(
    artifacts: list[str],
    environments: list[str],
    secrets: list[artifact_pb2.ArtifactStepSecret],
    script: str,
) -> artifact_pb2.ArtifactStep:
    """Build a bash step (Darwin). Matches Rust/Go/TS ``bash``."""
    # Deferred import breaks the artifact<->step cycle (artifact imports shell).
    from vorpal_sdk.artifact import get_env_key

    step_environments = [e for e in environments if not e.startswith("PATH=")]

    step_path_bins = ":".join(f"{get_env_key(a)}/bin" for a in artifacts)
    step_path = f"{step_path_bins}:{_DEFAULT_PATH}"

    for environment in environments:
        if environment.startswith("PATH="):
            path_value = environment[len("PATH=") :]
            if path_value:
                step_path = f"{path_value}:{step_path}"

    step_environments.append("HOME=$VORPAL_WORKSPACE")
    step_environments.append(f"PATH={step_path}")

    step_script = f"#!/bin/bash\nset -euo pipefail\n\n{script}\n"

    return artifact_pb2.ArtifactStep(
        entrypoint="bash",
        script=step_script,
        secrets=_dedupe_secrets(secrets),
        arguments=[],
        artifacts=artifacts,
        environments=step_environments,
    )


def bwrap(
    arguments: list[str],
    artifacts: list[str],
    environments: list[str],
    rootfs: str | None,
    secrets: list[artifact_pb2.ArtifactStepSecret],
    script: str,
) -> artifact_pb2.ArtifactStep:
    """Build a bwrap step (Linux). Matches Rust/Go/TS ``bwrap``.

    CRITICAL: argument-list ordering must be identical to Rust/Go/TS.
    """
    from vorpal_sdk.artifact import get_env_key

    step_arguments: list[str] = [
        "--unshare-all",
        "--share-net",
        "--clearenv",
        "--chdir",
        "$VORPAL_WORKSPACE",
        "--gid",
        "1000",
        "--uid",
        "1000",
        "--dev",
        "/dev",
        "--proc",
        "/proc",
        "--tmpfs",
        "/tmp",
        "--bind",
        "$VORPAL_OUTPUT",
        "$VORPAL_OUTPUT",
        "--bind",
        "$VORPAL_WORKSPACE",
        "$VORPAL_WORKSPACE",
        "--setenv",
        "VORPAL_OUTPUT",
        "$VORPAL_OUTPUT",
        "--setenv",
        "VORPAL_WORKSPACE",
        "$VORPAL_WORKSPACE",
        "--setenv",
        "HOME",
        "$VORPAL_WORKSPACE",
    ]

    step_artifacts: list[str] = []

    if rootfs is not None:
        rootfs_env = get_env_key(rootfs)
        step_arguments.extend(
            [
                "--ro-bind",
                f"{rootfs_env}/bin",
                "/bin",
                "--ro-bind",
                f"{rootfs_env}/etc",
                "/etc",
                "--ro-bind",
                f"{rootfs_env}/lib",
                "/lib",
                "--ro-bind-try",
                f"{rootfs_env}/lib64",
                "/lib64",
                "--ro-bind",
                f"{rootfs_env}/sbin",
                "/sbin",
                "--ro-bind",
                f"{rootfs_env}/usr",
                "/usr",
            ]
        )
        step_artifacts.append(rootfs)

    step_artifacts.extend(artifacts)

    for artifact in step_artifacts:
        env_key = get_env_key(artifact)
        step_arguments.append("--ro-bind")
        step_arguments.append(env_key)
        step_arguments.append(env_key)
        step_arguments.append("--setenv")
        step_arguments.append(env_key.replace("$", ""))
        step_arguments.append(env_key)

    step_path_bins = ":".join(f"{get_env_key(a)}/bin" for a in step_artifacts)
    step_path = f"{step_path_bins}:{_DEFAULT_PATH}"

    for environment in environments:
        if environment.startswith("PATH="):
            path_value = environment[len("PATH=") :]
            if path_value:
                step_path = f"{path_value}:{step_path}"

    step_arguments.append("--setenv")
    step_arguments.append("PATH")
    step_arguments.append(step_path)

    for environment in environments:
        eq_idx = environment.find("=")
        key = environment[:eq_idx] if eq_idx != -1 else environment
        value = environment[eq_idx + 1 :] if eq_idx != -1 else ""
        if key.startswith("PATH"):
            continue
        step_arguments.append("--setenv")
        step_arguments.append(key)
        step_arguments.append(value)

    step_arguments.extend(arguments)

    step_script = f"#!/bin/bash\nset -euo pipefail\n\n{script}\n"

    return artifact_pb2.ArtifactStep(
        entrypoint="bwrap",
        script=step_script,
        secrets=_dedupe_secrets(secrets),
        arguments=step_arguments,
        artifacts=step_artifacts,
        environments=[f"PATH={_DEFAULT_PATH}"],
    )


def shell(
    context: BuildContext,
    artifacts: list[str],
    environments: list[str],
    script: str,
    secrets: list[artifact_pb2.ArtifactStepSecret],
) -> artifact_pb2.ArtifactStep:
    """Dispatch to ``bash`` on Darwin, ``bwrap`` on Linux.

    Matches Rust/Go/TS ``shell``. The Linux branch requires the
    ``linux-vorpal`` rootfs builder, deferred to a later phase.
    """
    step_system = context.get_system()

    if step_system in (
        artifact_pb2.AARCH64_DARWIN,
        artifact_pb2.X8664_DARWIN,
    ):
        return bash(artifacts, environments, secrets, script)

    if step_system in (
        artifact_pb2.AARCH64_LINUX,
        artifact_pb2.X8664_LINUX,
    ):
        raise NotImplementedError(
            "shell() on Linux targets requires the linux-vorpal rootfs builder, "
            "which lands in a later phase"
        )

    raise ValueError(f"unsupported system: {step_system}")


def docker(
    arguments: list[str],
    artifacts: list[str],
) -> artifact_pb2.ArtifactStep:
    """Build a docker step. Matches Rust/Go/TS ``docker``."""
    return artifact_pb2.ArtifactStep(
        entrypoint="docker",
        secrets=[],
        arguments=arguments,
        artifacts=artifacts,
        environments=[f"PATH={_DEFAULT_PATH}"],
    )
